magicctapipe/scripts/lst1_magic/lst1_magic_event_coincidence.py

```diff
@@ -41,514 +55,357 @@
+        # MAGIC events existing in the window, including the edges, are
+        # recognized as the coincident events. At first, we scan the
+        # number of coincident events in each time offset and find the
+        # offset maximizing the number of events. Then, we calculate the
+        # average offset weighted by the number of events around the
+        # maximizing offset. Finally, we again check the coincidence at
+        # the average offset and then save the coincident events.
+
+        # Note that there are two conditions for the event coincidence.
+        # The first one includes both edges of the coincidence window,
+        # and the other one includes only the right edge. The latter
+        # is used to estimate the number of coincident events between
+        # the time offset steps.
 
         n_events_lst = len(time_lst)
 
-        n_events_stereo = np.zeros(len(time_offsets), dtype=int)
-        n_events_stereo_btwn = np.zeros(len(time_offsets), dtype=int)
+        n_coincidence = np.zeros(len(time_offsets), dtype=int)
+        n_coincidence_btwn = np.zeros(len(time_offsets), dtype=int)
 
         logger.info("\nChecking the event coincidence...")
 
-        for i_step, offset in enumerate(time_offsets):
+        for i_step, time_offset in enumerate(time_offsets):
 
-            time_lolim = np.round(
-                time_lst + offset - window_width / 2, decimals=precision
-            )
-            time_uplim = np.round(
-                time_lst + offset + window_width / 2, decimals=precision
-            )
+            time_lolim = time_lst + time_offset - window_half_width
+            time_uplim = time_lst + time_offset + window_half_width
 
             for i_evt in range(n_events_lst):
 
-                # Check the coincidence including the both edges:
-                condition = np.logical_and(
-                    time_magic >= time_lolim[i_evt],
-                    time_magic <= time_uplim[i_evt],
-                )
+                cond_lolim = time_magic.value >= time_lolim[i_evt].value
+                cond_lolim_wo_eq = time_magic.value > time_lolim[i_evt].value
+                cond_uplim = time_magic.value <= time_uplim[i_evt].value
+
+                # Check the coincidence including the both edges
+                mask = np.logical_and(cond_lolim, cond_uplim)
 
-                if np.count_nonzero(condition) == 1:
-                    n_events_stereo[i_step] += 1
+                if np.count_nonzero(mask) == 1:
+                    n_coincidence[i_step] += 1
 
-                # Check the coincidence including only the right edge:
-                condition_btwn = np.logical_and(
-                    time_magic > time_lolim[i_evt],
-                    time_magic <= time_uplim[i_evt],
-                )
+                # Check the coincidence including only the right edge
+                mask_btwn = np.logical_and(cond_lolim_wo_eq, cond_uplim)
```

Comment:
why do you use here a different condition? shouldn't the difference between < and <== be very small?

Reply:
The second condition (including only the right edge) is used for estimating the number of coincident events between the 100 ns offset steps. Since the final digit of the timestamps is 100 ns, the number of coincident events between the steps should be constant. And it can be estimated from each 100 ns step position by removing the left edge condition, because the offset scan is performed from smaller to larger values, and the events coincident with the left edge condition are the ones becoming not coincident between the steps. Those are useful to estimate the number of coincident events at the average offset, which is usually between the offset steps.

But I realized that the checks are not needed at the first scan, and if we remove those checks we could speed up the algorithm. I will update the code now.

Reply:
Done. Now the speed becomes surprisingly faster by ~ 50%.